File: mobile/lib/core/utils/property_share.dart
import 'package:url_launcher/url_launcher.dart';
import 'property_nav.dart';

enum PropertyShareKind { rental, bnb }

class PropertyShare {
  static const String baseOrigin = 'https://rental.oweru.com';

  /// Share preview URL — WhatsApp crawls this for og:image property photo preview.
  static String buildUrl(int propertyId, {int? agentId, PropertyShareKind kind = PropertyShareKind.rental}) {
    final segment = kind == PropertyShareKind.bnb ? 'bnb' : 'property';
    final q = kind == PropertyShareKind.rental && agentId != null ? '?agent=$agentId' : '';
    return '$baseOrigin/api/public/share/$segment/$propertyId$q';
  }

  static String buildPageUrl(int propertyId, {int? agentId, PropertyShareKind kind = PropertyShareKind.rental}) {
    if (kind == PropertyShareKind.bnb) return '$baseOrigin/bnb/$propertyId';
    final q = agentId != null ? '?agent=$agentId' : '';
    return '$baseOrigin/property/$propertyId$q';
  }

  static PropertyShareKind kindFor(Map<String, dynamic> property) {
    if (PropertyNav.isBnbListing(property)) return PropertyShareKind.bnb;
    return PropertyShareKind.rental;
  }

  static String buildMessage(String title, String url) {
    final name = title.trim().isEmpty ? 'this property' : title.trim();
    return 'Check out this property on Oweru: $name\n$url';
  }

  static Future<void> shareWhatsApp(String message) async {
    final uri = Uri.parse('https://wa.me/?text=${Uri.encodeComponent(message)}');
    if (await canLaunchUrl(uri)) {
      await launchUrl(uri, mode: LaunchMode.externalApplication);
    }
  }
}

File: frontend/src/utils/propertyShare.ts
const DEFAULT_ORIGIN = 'https://rental.oweru.com';

export function buildPropertyShareUrl(
  propertyId: number | string,
  agentId?: number | string | null,
): string {
  const base = typeof window !== 'undefined' ? window.location.origin : DEFAULT_ORIGIN;
  const q = agentId != null && `${agentId}`.trim() !== '' ? `?agent=${agentId}` : '';
  return `${base}/property/${propertyId}${q}`;
}

export function buildPropertyShareMessage(title: string, url: string): string {
  const name = title?.trim() || 'this property';
  return `Check out this property on Oweru: ${name}\n${url}`;
}

export function openWhatsAppShare(message: string): void {
  window.open(`https://wa.me/?text=${encodeURIComponent(message)}`, '_blank', 'noopener,noreferrer');
}
